docs/microbenchmark-design-guidelines.md

```diff
@@ -181,7 +181,7 @@ If you want to get a better understanding of it, you should read [this blog post
 
 ### OperationsPerInvoke
 
-`[GlobalSetup]` and `[IterationSetup]` might not be enough if you want to setup some nano-benchmarks.
+You also need to make sure that the benchmark itself performs enough of computations for a single invocation to run longer than 100ms. **If you don't, your benchmark will be entirely invalid.**
```

Comment:
@adamsitnik this comment "You also need to make sure that the benchmark itself performs enough of computations for a single invocation to run longer than 100ms. If you don't, your benchmark will be entirely invalid." was already in this file. With your guidance to not add a loop to reach 100ms, I realize I don't understand the comment. Can you clarify? And probably we can then remove this line.

Reply:
@danmoseley BenchmarkDotNet by default scales the number of invocation per iteration. Example: if operation takes 1 ms and the `IterationTime` is configured to 250 ms, it's going to choose a number close to 250. That is why most of the users should never put any kinds of loops into their benchmarks

* `[GlobalSetup]` and `[GlobalCleanup]` are performed just once and they don't affect scaling
* `[IterationSetup]` and `[IterationCleanup]` are performed once before every iteration. Since most of the users want to setup iteration because the benchmark needs a setup for every invocation, when these attributes are used BDN performs a single invocation per iteration. Since `Stopwatch` has a limited resolution, BenchmarkDotNet emits a warning if `[IterationSetup]` has been used but the iteration lasted less than 100 ms. This is why if we use `[IterationSetup]` we need to ensure that the work it performs must last for at least 100 ms.
* `OperationsPerInvoke` - scaling is not affected, it was designed to amortize small (nanoseconds) setup cost amongst multiple operations. In the example below, the cost of creating a span (calling span ctor) is amortized amongst all slice operations. The reported time is `ctor_time / OperationsPerInvoke  + slice_time = ctor_time / 16 + slice_time`. 

```cs
[Benchmark(OperationsPerInvoke = 16)]
public Span<byte> Slice()
{
    System.Span<byte> span = new System.Span<T>(_nonEmptyArray); // amortized
    span = span.Slice(1); span = span.Slice(1); span = span.Slice(1); span = span.Slice(1);
    span = span.Slice(1); span = span.Slice(1); span = span.Slice(1); span = span.Slice(1);
    span = span.Slice(1); span = span.Slice(1); span = span.Slice(1); span = span.Slice(1);
    span = span.Slice(1); span = span.Slice(1); span = span.Slice(1); span = span.Slice(1);
    return span;
}
```

Please keep in mind that in the example above I've not used a loop, as it would introduce non-trivial cost compared the benchmarked operation (slicing) and could also potentially introduce alignment issues.